src/effects/VST3/VST3Effect.cpp

```diff
@@ -1217,3 +1183,128 @@ void VST3Effect::ReloadUserOptions()
 
    SetBlockSize(mUserBlockSize);
 }
+
+
+bool VST3Effect::TransferDataToWindow(const EffectSettings& settings)
+{
+   // Update parameter values in the plugin
+   if (StoreSettings(GetSettings(settings)))
+   {
+      return true;
+   }
+   
+   return false;
+}
+
+
+
+bool VST3Wrapper::ForEachParameter(ParameterVisitor visitor) const
```

Comment:
In the commit that introduces this function, don't just define it, but also insert uses of it in all possible places.

`getParameterCount` should be called directly in this function only.

I count four functions to rewrite in VST3Effect.cpp and one more in VST3Utils.cpp.  There is a function `FillParameters` which could be changed to take a reference to a wrapper instead of EditController.

This also implies the VST3Wrapper struct should be declared and defined in VST3Utils.h/.cpp, so that there is not a dependency cycle between those and VST3Effect.

Reply:
Recall also: it may be better to fill a vector of ParameterInfo to iterate, only once, say in the contructor of `VST3Wrapper`.  Then the for-each function just iterates the vector.

When the Instance class is introduced, construct its wrapper to share -- not reinitialize -- the vector of ParameterInfo that is in the EffectPlugin.  That would imply other changes in the constructor.

Reply:
Unfortunately I can not fill the vector of ParameterInfo in the VST3Wrapper constructor, because I need to reference `mEditController `and that gets ready only at the end of `VST3Effect `constructor. So I had to opt for an ::`Init `method instead. Later I could try to rework the `VST3Effect `contsructor, to move the stuff which initializes `mEditController `to the Wrapper.

Reply:
Then, some kind of create-once-on-demand function in the new base class can return the vector of ParameterInfo, and that function could fail if called to early, and the for-each function should check for that error and pass it along.